// @backstage/plugin-bazaar-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendCommon = require('@backstage/backend-common');
var backendPluginApi = require('@backstage/backend-plugin-api');
var router = require('./cjs/router-fd8ba920.cjs.js');
require('express');
require('express-promise-router');

var alpha = backendPluginApi.createBackendPlugin({
  pluginId: "bazaar",
  register(env) {
    env.registerInit({
      deps: {
        config: backendPluginApi.coreServices.rootConfig,
        database: backendPluginApi.coreServices.database,
        identity: backendPluginApi.coreServices.identity,
        logger: backendPluginApi.coreServices.logger,
        httpRouter: backendPluginApi.coreServices.httpRouter
      },
      async init({ database, config, identity, logger, httpRouter }) {
        httpRouter.use(
          await router.createRouter({
            database,
            config,
            identity,
            logger: backendCommon.loggerToWinstonLogger(logger)
          })
        );
      }
    });
  }
});

exports["default"] = alpha;
