// @backstage/plugin-app-backend/dist/alpha.cjs.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var backendPluginApi = require('@backstage/backend-plugin-api');
var router = require('./cjs/router-c47f2ed4.cjs.js');
var backendCommon = require('@backstage/backend-common');
var pluginAppNode = require('@backstage/plugin-app-node');
require('helmet');
require('express');
require('express-promise-router');
require('fs-extra');
require('path');
require('@backstage/config-loader');
require('luxon');
require('lodash/partition');
require('globby');

const appPlugin = backendPluginApi.createBackendPlugin({
  pluginId: "app",
  register(env) {
    let staticFallbackHandler;
    let schema;
    env.registerExtensionPoint(pluginAppNode.staticFallbackHandlerExtensionPoint, {
      setStaticFallbackHandler(handler) {
        if (staticFallbackHandler) {
          throw new Error(
            "Attempted to install a static fallback handler for the app-backend twice"
          );
        }
        staticFallbackHandler = handler;
      }
    });
    env.registerExtensionPoint(pluginAppNode.configSchemaExtensionPoint, {
      setConfigSchema(configSchema) {
        if (schema) {
          throw new Error(
            "Attempted to set config schema for the app-backend twice"
          );
        }
        schema = configSchema;
      }
    });
    env.registerInit({
      deps: {
        logger: backendPluginApi.coreServices.logger,
        config: backendPluginApi.coreServices.rootConfig,
        database: backendPluginApi.coreServices.database,
        httpRouter: backendPluginApi.coreServices.httpRouter
      },
      async init({ logger, config, database, httpRouter }) {
        var _a;
        const appPackageName = (_a = config.getOptionalString("app.packageName")) != null ? _a : "app";
        const winstonLogger = backendCommon.loggerToWinstonLogger(logger);
        const router$1 = await router.createRouter({
          logger: winstonLogger,
          config,
          database,
          appPackageName,
          staticFallbackHandler,
          schema
        });
        httpRouter.use(router$1);
      }
    });
  }
});

exports["default"] = appPlugin;
