// @swagger-api/apidom-reference/es/resolve/strategies/apidom/index.mjs
import stampit from 'stampit';
import { isElement, visit, cloneDeep } from '@swagger-api/apidom-core';
import ResolveStrategy from "../ResolveStrategy.mjs";
import ReferenceSet from "../../../ReferenceSet.mjs";
import Reference from "../../../Reference.mjs";
import { merge as mergeOptions } from "../../../options/util.mjs";
import ApiDOMResolveVisitor from "./visitor.mjs"; // @ts-ignore
const visitAsync = visit[Symbol.for('nodejs.util.promisify.custom')];
const ApiDOMResolveStrategy = stampit(ResolveStrategy, {
  init() {
    this.name = 'apidom';
  },
  methods: {
    canResolve(file) {
      var _file$parseResult;
      return file.mediaType.startsWith('application/vnd.apidom') && isElement((_file$parseResult = file.parseResult) === null || _file$parseResult === void 0 ? void 0 : _file$parseResult.result);
    },
    async resolve(file, options) {
      var _options$resolve$stra;
      const referenceValue = (_options$resolve$stra = options.resolve.strategyOpts.apidom) !== null && _options$resolve$stra !== void 0 && _options$resolve$stra.clone ? cloneDeep(file.parseResult) : file.parseResult;
      const reference = Reference({
        uri: file.uri,
        value: referenceValue
      });
      const mergedOptions = mergeOptions(options, {
        resolve: {
          internal: false
        }
      });
      const visitor = ApiDOMResolveVisitor({
        reference,
        options: mergedOptions
      });
      const refSet = ReferenceSet();
      refSet.add(reference);
      await visitAsync(refSet.rootRef.value, visitor);
      return refSet;
    }
  }
});
export default ApiDOMResolveStrategy;
