// @backstage/plugin-home/dist/alpha.esm.js
import React from 'react';
import { createRouteRef, createExtensionDataRef, createPageExtension, createExtensionInput, coreExtensionData, createPlugin } from '@backstage/frontend-plugin-api';
import { compatWrapper } from '@backstage/core-compat-api';

const rootRouteRef = createRouteRef();
const titleExtensionDataRef = createExtensionDataRef("title");
const homePage = createPageExtension({
  defaultPath: "/home",
  routeRef: rootRouteRef,
  inputs: {
    props: createExtensionInput(
      {
        children: coreExtensionData.reactElement.optional(),
        title: titleExtensionDataRef.optional()
      },
      {
        singleton: true,
        optional: true
      }
    )
  },
  loader: ({ inputs }) => import('./esm/index-abfba912.esm.js').then(
    (m) => {
      var _a, _b;
      return compatWrapper(
        /* @__PURE__ */ React.createElement(
          m.HomepageCompositionRoot,
          {
            children: (_a = inputs.props) == null ? void 0 : _a.output.children,
            title: (_b = inputs.props) == null ? void 0 : _b.output.title
          }
        )
      );
    }
  )
});
var alpha = createPlugin({
  id: "home",
  extensions: [homePage]
});

export { alpha as default, titleExtensionDataRef };
